dask/array/creation.py

```diff
@@ -501,6 +501,81 @@ def diag(v):
     return Array(graph, name, (chunks_1d, chunks_1d), dtype=v.dtype)
 
 
+@wraps(np.diagonal)
+def diagonal(a, offset=0, axis1=0, axis2=1):
+    name = 'diagonal-' + tokenize(a, offset, axis1, axis2)
+
+    if a.ndim < 2:
+        # NumPy uses `diag` as we do here.
+        raise ValueError("diag requires an array of at least two dimensions")
+
+    if axis1 == axis2:
+        raise ValueError("axis1 and axis2 cannot be the same")
+
+    if offset >= 0:
+        dim1_offset = 0
+        dim2_offset = +offset
+    else:
+        dim1_offset = -offset
+        dim2_offset = 0
+
+    if axis1 > axis2:
+        axis1, axis2 = axis2, axis1
+        dim1_offset, dim2_offset = dim2_offset, dim1_offset
+        offset *= -1
+
+    left_idx = list(range(len(a.shape)))
+    del left_idx[axis2]
+    del left_idx[axis1]
+
+    left_shape = tuple([a.shape[i] for i in left_idx])
+    right_shape = (max(min(a.shape[axis1] - dim1_offset,
+                           a.shape[axis2] - dim2_offset), 0),)
+    shape = left_shape + right_shape
+
+    if isinstance(a, np.ndarray):
+
+        chunks = (1,) * len(left_shape) + (right_shape,)
+        chunks = normalize_chunks(chunks=chunks, shape=shape)
+
+        dsk = {}
+        a_indices = [range(i) for i in a.shape]
+        a_indices[axis1] = [slice(None)]
+        a_indices[axis2] = [slice(None)]
+        sub_shape = (a.shape[axis1], a.shape[axis2]) + (1,) * (len(shape) - 1)
+        for ai in product(*a_indices):
+            ci = tuple(i for i in ai if not isinstance(i, slice))
+            dsk[(name,) + ci + (0,)] = (np.diagonal, a[ai].reshape(sub_shape),
+                                        offset)
+
+        return Array(dsk, name, chunks, dtype=a.dtype)
+
+    if not isinstance(a, Array):
+        raise TypeError("a must be a dask array or numpy array, "
+                        "got {0}".format(type(a)))
+
+    chunks = list(a.chunks)
+    chunks[axis1] = -1
+    chunks[axis2] = -1
+    a = a.rechunk(chunks)
+
+    a_indices = [range(len(i)) for i in a.chunks]
+
+    def _diag(x, offset, axis1, axis2):
+        return np.diagonal(asarray(x), axis1=axis1, axis2=axis2,
+                           offset=offset)
```

Comment:
Sorry, one small question. Would it be possible to just use `np.diagonal` directly?

Reply:
`np.diag` is supposed to work with 1- or 2-d arrays only.

Reply:
Not sure I follow. Where is `np.diag` being used? My understanding was we are just wrapping `np.diagonal`, right? So am trying to understand why that is needed. Thoughts?

Reply:
This looks like a github bug? I'ave removed those lines being mentioned...